File: C/rip/rip.c
#define _CRT_SECURE_NO_WARNINGS

#include <stdio.h>
#include <string.h>

#define NODE_NUMBER 7
#define INFINITY 1024
#define INFO_SIZE sizeof(RoutingInfo)

typedef struct
{
	int destination;
	int cost;
	int nextHop;
} RoutingInfo;

RoutingInfo routingTable[NODE_NUMBER][NODE_NUMBER];
int ifChanged = 1;

void initRoutingTable(FILE *fp)
{
	int token, sn;
	sn = fscanf(fp, "%d", &token);
	for (int i = 0; i < NODE_NUMBER; i++)
	{
		for (int j = 0; j < NODE_NUMBER; j++)
		{
			if (sn != 0)
			{
				routingTable[i][j].cost = token;
			}
			else
			{
				char buffer;
				fscanf(fp, "%c", &buffer);
				if (buffer == 'i')
				{
					routingTable[i][j].cost = INFINITY;
				}
			}
			routingTable[i][j].destination = j;
			routingTable[i][j].nextHop = routingTable[i][j].cost == INFINITY || routingTable[i][j].cost == 0 ? NODE_NUMBER : j;
			sn = fscanf(fp, "%d", &token);
		}
	}
}

void receiveRoutingTable(int senderNum, int receiverNum, int senderInfo[])
{
	for (int i = 0; i < NODE_NUMBER; i++)
	{
		if (routingTable[receiverNum][i].cost > routingTable[receiverNum][senderNum].cost + senderInfo[i])
		{
			routingTable[receiverNum][i].cost = routingTable[receiverNum][senderNum].cost + senderInfo[i];
			routingTable[receiverNum][i].nextHop = senderNum;
			ifChanged = 1;
		}
	}
}

int isAdjacent(int nodeNum1, int nodeNum2)
{
	for (int i = 0; i < NODE_NUMBER; i++)
	{
		if (routingTable[nodeNum1][i].nextHop == nodeNum2)
		{
			return 1;
		}
	}
	return 0;
}

void sendRoutingTable()
{
	int senderInfos[NODE_NUMBER][NODE_NUMBER];
	for (int i = 0; i < NODE_NUMBER; i++)
	{
		for (int j = 0; j < NODE_NUMBER; j++)
		{
			senderInfos[i][j] = routingTable[i][j].cost;
		}
	}
	for (int i = 0; i < NODE_NUMBER; i++)
	{
		for (int j = 0; j < NODE_NUMBER; j++)
		{
			if (isAdjacent(i, j))
			{
				receiveRoutingTable(i, j, senderInfos[i]);
			}
		}
	}
}

void main()
{
	FILE *fp = fopen("test.txt", "r");
	initRoutingTable(fp);
	fclose(fp);
	while (ifChanged)
	{
		ifChanged = 0;
		sendRoutingTable();
		for (int i = 0; i < NODE_NUMBER; i++)
		{
			for (int j = 0; j < NODE_NUMBER; j++)
			{
				if (routingTable[i][j].cost == INFINITY)
				{
					printf("%c ", '*');
				}
				else
				{
					printf("%d ", routingTable[i][j].cost);
				}
			}
			printf("\n");
		}
		printf("--------------------------------------\n");
	}
}

File: Python/rip/rip.py
import os

INFINITY = 1024

class RoutingInfo():
    def __init__(self, destination, cost, nextHop):
        self.destination = destination
        self.cost = cost
        self.nextHop = nextHop

ifChanged = True

def initRoutingTable(f):
    nodeNumber = int(f.readline())
    routingTable = [[None] * nodeNumber for _ in range(nodeNumber)]
    for i in range(nodeNumber):
        line = f.readline().strip().split()
        for j in range(nodeNumber):
            try:
                token = int(line[j])
                routingTable[i][j] = RoutingInfo(j, token, 0)
            except ValueError:
                buffer = line[j]
                if buffer == "i":
                    routingTable[i][j] = RoutingInfo(j, INFINITY, 0)
            routingTable[i][j].nextHop = nodeNumber if routingTable[i][j].cost == INFINITY or routingTable[i][j].cost == 0 else j
    return routingTable

def isAdjacent(routingTable, nodeNum1, nodeNum2):
    for i in range(len(routingTable)):
        if routingTable[nodeNum1][i].nextHop == nodeNum2:
            return True
    return False

def receiveRoutingTable(routingTable, senderNum, receiverNum, senderInfo):
    global ifChanged
    for i in range(len(routingTable)):
        if routingTable[receiverNum][i].cost > routingTable[receiverNum][senderNum].cost + senderInfo[i]:
            routingTable[receiverNum][i].cost = routingTable[receiverNum][senderNum].cost + senderInfo[i]
            routingTable[receiverNum][i].nextHop = senderNum
            ifChanged = True

def sendRoutingTable(routingTable):
    nodeNumber = len(routingTable)
    senderInfos = [[None] * nodeNumber for _ in range(nodeNumber)]
    for i in range(nodeNumber):
        for j in range(nodeNumber):
            senderInfos[i][j] = routingTable[i][j].cost
    for i in range(nodeNumber):
        for j in range(nodeNumber):
            if isAdjacent(routingTable, i ,j):
                receiveRoutingTable(routingTable, i, j, senderInfos[i])

def main():
    if os.path.exists(r"Python\rip\test.txt"):
        routingTable = None
        with open(r"Python\rip\test.txt", "r") as f:
            routingTable = initRoutingTable(f)
        nodeNumber = len(routingTable)
        global ifChanged
        while ifChanged:
            ifChanged = False
            sendRoutingTable(routingTable)
            for routingInfos in routingTable:
                for i in range(nodeNumber):
                    if routingInfos[i].cost == INFINITY:
                        print("* ", end="")
                    else:
                        print(f"{routingInfos[i].cost} ", end="")
                print()
            print("------------------------------------")
    else:
        print("file not found")

if __name__ == "__main__":
    main()